### PR TITLE
CON-1127 Upgrade to latest Kotlin 1.7.20

### PR DESCRIPTION
**Pull request by filipesoliveira**

### CODE REVIEW

**Review by anien — APPROVED** [review]
Looks good.

**Comment by filipesoliveira** [comment]
**Note:**
image

**Comment by filipesoliveira** [comment]
SDK Build (Conclave Core):
- It fails when Kotlin is upgraded to 6.20 or 7.20
- It fails only on TC.
- ./gradlew test works fine on a local machine or an Azure machine
-

**Comment by shamsasari** [comment]
The error is occuring because `com.r3.conclave.enclave.Enclave$Companion.getMailDecryptingStream(Enclave.kt:73)` is trying to invoke the c'tor

`com.r3.conclave.mail.internal.MailDecryptingStream.<init>(Ljava/io/InputStream;Ljava/security/PrivateKey;ILkotlin/jvm/internal/DefaultConstructorMarker;)V`.

It shouldn't be doing this, because this c;tor doesn't exist. The c'tor that does exist is 

`(Ljava/io/InputStream;Ljava/security/PrivateKey;ILcom/r3/conclave/shaded/kotlin/jvm/internal/DefaultConstructorMarker;)V`

i.e. the difference is in the last parameter. The `DefaultConstructorMarker` should be our shaded one (`com.r3.conclave.shaded...`) and not the standard kotlin one.

The question is why isn't the shading process updating `Enclave$Companion.getMailDecryptingStream` to use our shaded c;tor.

**Comment by shamsasari** [comment]
Can you try `./gradlew -i conclave-enclave:shadowJar` and see if the shadow plugin prints out any debug logging. What we want to find out is why it is not rewriting the bytecode for `Enclave.Companion.getMailDecryptingStream` to use the shaded c;tor.

**Comment by filipesoliveira** [comment]
The integration tests are successful when using the Kotlin version 6.20.

**Comment by filipesoliveira** [comment]
Todo:
- Remove the generated enclave_jni.
- Fix the task clean to also delete the enclave_jni that is generated by the cmake.
- Update the dependensOn so the enclave_jni is created before it is needed.
- Undo the changes that are not required.
- Leave a comment about the fix done.

**Comment by shamsasari** [comment]
Huh, merged without approval! I guess it must have been Anna's previous approval?

File: cpp/build.gradle
Comment (shamsasari) [inline cpp/build.gradle:3]: typo: clean

File: build.gradle
Comment (shamsasari) [inline build.gradle:294]: What's the reason for this rename?

Comment (filipesoliveira) [reply build.gradle:294]: There was a name collision between the function `project`, in line 290 and the variable `project`. This causes in line 307 where the function `project` is required again.

**Comment by filipesoliveira** [comment]
PR https://github.com/R3Conclave/conclave-core-sdk/pull/69 addresses the suggestions made in this PR after it was merged.

File: build.gradle
Comment (shamsasari) [reply build.gradle:294]: I presumed it was due to groovy. Another reason to switch to the kotlin DSL.

**Pull request merged**
